### PR TITLE
Add Base Exception class

### PR DESCRIPTION
**Pull request by everly-gif**
Following the changes in this [PR](https://github.com/appwrite/appwrite/pull/4575), we need to add some base Database Exception that we could use to distinguish between our errors and native errors. This PR implements a base exception class.

### CODE REVIEW

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Looks good 👌  

Can you also scan through the code base and check if there are usages of exceptions other than Utopia\Database\Exception or Utopia\Database\Exception\Authorization etc ? 

Basically we want to ensure that every custom error thrown by the database library is an instance of the Database library exception or its sub classes.

**Comment by everly-gif** [comment]
> Looks good 👌
> 
> Can you also scan through the code base and check if there are usages of exceptions other than Utopia\Database\Exception or Utopia\Database\Exception\Authorization etc ?
> 
> Basically we want to ensure that every custom error thrown by the database library is an instance of the Database library exception or its sub classes.

I have updated every instance where we throw an exception to be a type of `Utopia\Database\Exception`. But there are still certain uses of `\Exception` such as places below

```
  try  {
            new \DateTime($datetime);
        }
        catch(\Exception $e) {
            return false;
        }
```
ie. in cases where we don't throw an exception explicitly.

File: src/Database/Adapter/MariaDB.php [hunk 1]
```diff
@@ -551,7 +552,7 @@ public function createDocument(string $collection, Document $document): Document
         }
 
         if (!$this->getPDO()->commit()) {
-            throw new Exception('Failed to commit transaction');
+            throw new DatabaseException('Failed to commit transaction');
```
Comment (christyjacob4) [inline src/Database/Adapter/MariaDB.php:555]: Let's hide this. We don't need to show this on the UI

File: src/Database/Adapter/MariaDB.php [hunk 2]
```diff
@@ -767,15 +768,15 @@ public function deleteDocument(string $collection, string $id): bool
...
+            $stmt->execute() || throw new DatabaseException('Failed to delete document');
+            $stmtPermissions->execute() || throw new DatabaseException('Failed to clean permissions');
         } catch (\Throwable $th) {
             $this->getPDO()->rollBack();
             throw new Exception($th->getMessage());
         }
 
         if (!$this->getPDO()->commit()) {
-            throw new Exception('Failed to commit transaction');
+            throw new DatabaseException('Failed to commit transaction');
```
Comment (christyjacob4) [inline src/Database/Adapter/MariaDB.php:779]: Let's hide this. We don't need to show this on the UI

File: src/Database/Adapter/MariaDB.php [hunk 3]
```diff
@@ -1311,7 +1312,7 @@ public function getAttributeWidth(Document $collection): int
                     $total += 19; // 2022-06-26 14:46:24
                     break;
                 default:
-                    throw new Exception('Unknown Type');
+                    throw new DatabaseException('Unknown Type');
```
Comment (christyjacob4) [inline src/Database/Adapter/MariaDB.php:1315]:
```suggestion
                    throw new DatabaseException('Unknown Type. Must be one of ${Database::VAR_BOOLEAN}, ...');
```

Can we format this error message to be more descriptive. Let's add all the types in the message

File: src/Database/Adapter/MariaDB.php [hunk 4]
```diff
@@ -1666,7 +1667,7 @@ protected function getSQLType(string $type, int $size, bool $signed = true): str
                 return 'DATETIME(3)';
                 break;
             default:
-                throw new Exception('Unknown Type');
+                throw new DatabaseException('Unknown Type');
```
Comment (christyjacob4) [inline src/Database/Adapter/MariaDB.php:1670]: Same as the comment above

File: src/Database/Adapter/MariaDB.php [hunk 5]
```diff
@@ -1731,7 +1732,7 @@ protected function getSQLOperator(string $method): string
                 return '>=';
 
             default:
-                throw new Exception('Unknown method:' . $method);
+                throw new DatabaseException('Unknown method:' . $method);
```
Comment (christyjacob4) [inline src/Database/Adapter/MariaDB.php:1735]: Let's make the message more descriptive like my comment above

File: src/Database/Adapter/MariaDB.php [hunk 6]
```diff
@@ -1757,7 +1758,7 @@ protected function getSQLIndexType(string $type): string
                 return 'FULLTEXT INDEX';
 
             default:
-                throw new Exception('Unknown Index Type:' . $type);
+                throw new DatabaseException('Unknown Index Type:' . $type);
```
Comment (christyjacob4) [inline src/Database/Adapter/MariaDB.php:1761]: Same as above

File: src/Database/Adapter/MariaDB.php [hunk 7]
```diff
@@ -1788,7 +1789,7 @@ protected function getSQLIndex(string $collection, string $id,  string $type, ar
                 break;
 
             default:
-                throw new Exception('Unknown Index Type:' . $type);
+                throw new DatabaseException('Unknown Index Type:' . $type);
```
Comment (christyjacob4) [inline src/Database/Adapter/MariaDB.php:1792]: Same as above

File: src/Database/Adapter/Mongo.php [hunk 8]
```diff
@@ -1008,7 +1010,7 @@ protected function getQueryOperator(string $operator): string
                 return '$search';
 
             default:
-                throw new Exception('Unknown Operator:' . $operator);
+                throw new DatabaseException('Unknown Operator:' . $operator);
```
Comment (christyjacob4) [inline src/Database/Adapter/Mongo.php:1013]:
```suggestion
                throw new DatabaseException('Unknown Operator:' . $operator . "Must be one of ..... ");
```
Let's make this message more descriptive

File: src/Database/Adapter/Mongo.php [hunk 9]
```diff
@@ -1025,7 +1027,7 @@ protected function getOrder(string $order): int
         return match ($order) {
             Database::ORDER_ASC => 1,
             Database::ORDER_DESC => -1,
-            default => throw new Exception('Unknown sort order:' . $order),
+            default => throw new DatabaseException('Unknown sort order:' . $order),
```
Comment (christyjacob4) [inline src/Database/Adapter/Mongo.php:1030]: same as above

File: src/Database/Adapter/Mongo/MongoClient.php [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+
+namespace Utopia\Database\Adapter\Mongo;
+
+use Exception;
+use MongoDB\BSON;
+use Swoole\Client;
+use Swoole\Coroutine\Client as CoroutineClient;
+use Utopia\Database\Adapter\Mongo\Auth;
+use Utopia\Database\Adapter\Mongo\MongoClientOptions;
+use Utopia\Database\Document;
```
Comment (christyjacob4) [inline src/Database/Adapter/Mongo/MongoClient.php:11]: Can you sync your branch with the base branch ? I don't think this file is expected to be here.

File: src/Database/Adapter/MySQL.php [hunk 11]
```diff
@@ -41,7 +41,7 @@ protected function getSQLIndex(string $collection, string $id, string $type, arr
                 break;
 
             default:
-                throw new Exception('Unknown Index Type:' . $type);
+                throw new DatabaseException('Unknown Index Type:' . $type);
```
Comment (christyjacob4) [inline src/Database/Adapter/MySQL.php:44]: Let's have a more descriptive error message

File: src/Database/Adapter/SQLite.php [hunk 12]
```diff
@@ -367,15 +368,15 @@ public function createDocument(string $collection, Document $document): Document
...
+                    throw new Duplicate('Duplicated document');
                 break;
                 default:
                     throw $e;
             }
         }
 
         if (!$this->getPDO()->commit()) {
-            throw new Exception('Failed to commit transaction');
+            throw new DatabaseException('Failed to commit transaction');
```
Comment (christyjacob4) [inline src/Database/Adapter/SQLite.php:379]: Let's hide this. We don't need to show this on the UI

File: src/Database/Adapter/SQLite.php [hunk 13]
```diff
@@ -563,7 +564,7 @@ public function updateDocument(string $collection, Document $document): Document
...
```
Comment (christyjacob4) [inline src/Database/Adapter/SQLite.php:567]: Let's hide this. We don't need to show this on the UI

File: src/Database/Adapter/SQLite.php [hunk 14]
```diff
@@ -607,7 +608,7 @@ protected function getSQLIndexType(string $type): string
                 return 'UNIQUE INDEX';
 
             default:
-                throw new Exception('Unknown Index Type:' . $type);
+                throw new DatabaseException('Unknown Index Type:' . $type);
```
Comment (christyjacob4) [inline src/Database/Adapter/SQLite.php:611]: We need a more descriptive error message like my comments above

File: src/Database/Adapter/SQLite.php [hunk 15]
```diff
@@ -638,7 +639,7 @@ protected function getSQLIndex(string $collection, string $id,  string $type, ar
                 break;
 
             default:
-                throw new Exception('Unknown Index Type:' . $type);
+                throw new DatabaseException('Unknown Index Type:' . $type);
```
Comment (christyjacob4) [inline src/Database/Adapter/SQLite.php:642]: Same as above

File: src/Database/Database.php [hunk 16]
```diff
@@ -703,29 +704,29 @@ public function createAttribute(string $collection, string $id, string $type, in
...
+                    throw new DatabaseException('Max size allowed for int is: ' . number_format($limit));
                 }
                 break;
             case self::VAR_FLOAT:
             case self::VAR_BOOLEAN:
             case self::VAR_DATETIME:
                 break;
             default:
-                throw new Exception('Unknown attribute type: ' . $type);
+                throw new DatabaseException('Unknown attribute type: ' . $type);
```
Comment (christyjacob4) [inline src/Database/Database.php:722]: More descriptive error message

File: src/Database/Database.php [hunk 17]
```diff
@@ -795,16 +796,16 @@ protected function validateDefaultTypes(string $type, mixed $default): void
...
                 break;
             case self::VAR_DATETIME:
                 if ($defaultType !== self::VAR_STRING) {
-                    throw new Exception('Default value ' . $default . ' does not match given type ' . $type);
+                    throw new DatabaseException('Default value ' . $default . ' does not match given type ' . $type);
                 }
                 break;
             default:
-                throw new Exception('Unknown attribute type: ' . $type);
+                throw new DatabaseException('Unknown attribute type: ' . $type);
```
Comment (christyjacob4) [inline src/Database/Database.php:808]: More descriptive error message

File: src/Database/Database.php [hunk 18]
```diff
@@ -961,22 +962,22 @@ public function updateAttribute(string $collection, string $id, string $type = n
...
+                            throw new DatabaseException('Max size allowed for int is: ' . number_format($limit));
                         }
                         break;
                     case self::VAR_FLOAT:
                     case self::VAR_BOOLEAN:
                     case self::VAR_DATETIME:
                         break;
                     default:
-                        throw new Exception('Unknown attribute type: ' . $type);
+                        throw new DatabaseException('Unknown attribute type: ' . $type);
```
Comment (christyjacob4) [inline src/Database/Database.php:980]: More descriptive error message

File: src/Database/Database.php [hunk 19]
```diff
@@ -1219,24 +1220,24 @@ public function createIndex(string $collection, string $id, string $type, array
...
             case self::INDEX_FULLTEXT:
                 if (!$this->adapter->getSupportForUniqueIndex()) {
-                    throw new Exception('Fulltext index is not supported');
+                    throw new DatabaseException('Fulltext index is not supported');
                 }
                 break;
 
             default:
-                throw new Exception('Unknown index type: ' . $type);
+                throw new DatabaseException('Unknown index type: ' . $type);
```
Comment (christyjacob4) [inline src/Database/Database.php:1240]: More descriptive error message

File: src/Database/Validator/Roles.php [hunk 20]
```diff
@@ -177,7 +178,7 @@ public function isValid($roles): bool
 
             try {
                 $role = Role::parse($role);
-            } catch (\Exception $e) {
+            } catch (Exception $e) {
```
Comment (christyjacob4) [inline src/Database/Validator/Roles.php:181]: Hmm.. is this necessary ? Do we need to explicitly import the native exception class ?

File: src/Database/Role.php [hunk 21]
```diff
@@ -2,6 +2,8 @@
 
 namespace Utopia\Database;
 
+use Utopia\Database\Exception as DatabaseException;
```
Comment (christyjacob4) [inline src/Database/Role.php:5]: I don't think we need to expose these errors. Let's revert this file as its in the `main` branch

File: src/Database/Query.php [hunk 22]
```diff
@@ -171,7 +172,7 @@ public static function parse(string $filter): self
 
         // Check for deprecated query syntax
         if (\str_contains($method, '.')) {
-            throw new \Exception("Invalid query method");
+            throw new DatabaseException("Invalid query method");
```
Comment (christyjacob4) [inline src/Database/Query.php:175]: Okay after reviewing Role, Query and Permission I don't think we need to expose any of those errors in the UI either. Let's revert these files back too

File: src/Database/Adapter/Mongo/MongoClient.php (hunk 10 above)
Comment (everly-gif) [reply src/Database/Adapter/Mongo/MongoClient.php:11]: I am on the latest version of `main`

File: src/Database/Adapter/MariaDB.php [hunk 23]
```diff
@@ -1311,7 +1312,7 @@ public function getAttributeWidth(Document $collection): int
                     $total += 19; // 2022-06-26 14:46:24
                     break;
                 default:
-                    throw new Exception('Unknown Type');
+                    throw new DatabaseException('Unknown Type. Must be one of ${Database::VAR_INTEGER}, ${Database::VAR_FLOAT}, ${Database::VAR_BOOLEAN}, ${Database::VAR_DOCUMENT}, ${Database::VAR_DATETIME}');
```
Comment (christyjacob4) [inline src/Database/Adapter/MariaDB.php:1315]:
```suggestion
                    throw new DatabaseException("Unknown Type. Must be one of ${Database::VAR_INTEGER}, ${Database::VAR_FLOAT}, ${Database::VAR_BOOLEAN}, ${Database::VAR_DOCUMENT}, ${Database::VAR_DATETIME}");
```

String interpolation only works when the string is enclosed in double quotes

File: src/Database/Adapter/Mongo.php [hunk 24]
```diff
@@ -1008,7 +1010,7 @@ protected function getQueryOperator(string $operator): string
                 return '$search';
 
             default:
-                throw new Exception('Unknown Operator:' . $operator);
+                throw new DatabaseException('Unknown Operator:' . $operator .'. Must be one of ${Query::TYPE_EQUAL}, ${Query::NOTEQUAL, ${Query::TYPE_LESSER}, ${Query::TYPE_LESSEREQUAL}, ${Query::TYPE_GREATER}, ${Query::TYPE_GREATEREQUAL}, ${Query::TYPE_CONTAINS}, {Query::TYPE_SEARCH}');
```
Comment (christyjacob4) [inline src/Database/Adapter/Mongo.php:1013]:
```suggestion
                throw new DatabaseException('Unknown Operator:' . $operator .". Must be one of ${Query::TYPE_EQUAL}, ${Query::NOTEQUAL}, ${Query::TYPE_LESSER}, ${Query::TYPE_LESSEREQUAL}, ${Query::TYPE_GREATER}, ${Query::TYPE_GREATEREQUAL}, ${Query::TYPE_CONTAINS}, {Query::TYPE_SEARCH}");
```

File: src/Database/Adapter/Mongo/MongoClient.php (hunk 10 above)
Comment (everly-gif) [reply src/Database/Adapter/Mongo/MongoClient.php]: removed this file

**Review by eldadfux — APPROVED** [review]

**Pull request merged**